rfd/0203-database-healthchecks.md

```diff
@@ -0,0 +1,606 @@
+The maximum time is bounded by the health check timeout, which is itself bounded to be less than the interval.
+For example, (Interval=10s, Timeout=10s):
+
+1. health checker starts
+2. check times out after 10s - startTime=0s, endTime=10s
+3. check times out after 10s - startTime=10s, endTime=20s
+4. check times out after 10s - startTime=20s, endTime=30s
+
+Health checks will run periodically as long as the database remains registered with the DB agent.
+
+The health checker will maintain a single counter tracking the number of consecutive passing or failing checks.
+If the last check passed but the current check fails, then the counter resets to 1 and tracks the number of consecutive failing checks.
+Likewise, if the last check failed but the current check passes, then the counter resets to 1.
+
+If the number of consecutive passing or failing checks reaches the healthy or unhealthy threshold, respectively, then the health status is transitioned to the corresponding status: `healthy` or `unhealthy`.
+
+Example behavior with the default settings (Interval=10s, Timeout=5s, HealthyThreshold=2, UnhealthyThreshold=1):
+
+1. Health checker starts - status=`init`, startTime=0s, endTime=0s, count=0, lastErr=nil
+2. health check passes   - status=`init`, startTime=0s, endTime=0s, count=1, lastErr=nil
+3. health check fails    - status=`unhealthy`, startTime=10s, endTime=15s, count=1, lastErr="connection timeout"
+4. health check passes   - status=`unhealthy`, startTime=20s, endTime=20s, count=1, lastErr=nil
+5. health check passes   - status=`healthy`, startTime=30s, endTime=30s, count=2, lastErr=nil
+
+### Proxy behavior
+
+The proxy currently selects HA agents randomly for user connections.
+
+That will be changed to group the agents by health status, shuffle each group, and then combine the groups to prioritize healthy over unhealthy agent connections.
+
+The priority of health statuses will be:
+1. `healthy`
+2. `init`
+3. `""` (unknown)
+4. `unhealthy`
+
+The justification for the `healthy` and `unhealthy` status relative ordering should be obvious.
+It is perhaps less obvious why `init` should be preferred over `""`.
+By definition, an `init` status represents zero failing checks and zero or more passing health checks, whereas `""` represents no health information at all.
+Therefore, the proxy should prefer `init` over `""` health status.
+
+### Security
+
+Only users who can `tctl get db_server` can see health info, so it's already guarded by RBAC.
+
+We will enforce a minimum interval between health checks to prevent accidentally or intentionally dialing too often from agent to database.
+
+### Privacy
+
+N/A
+
+### Proto Specification
+
+A new message, HealthCheckSpec, will be added as field named `health_check` in DatabaseSpecV3 and ClusterNetworkingConfigSpecV2.
```

Comment:
In this approach, the DB agent attempts to modify the DatabaseV3 object. However, in the current database architecture, the `types.RoleDatabase` has only `services.RO()` permission on the `types.KindDatabase (DatabaseV3)` object.

See reference: [Teleport permissions](https://github.com/gravitational/teleport/blob/master/lib/authz/permissions.go#L1079).

This was implemented to resolve a security issue in [PR #25247](https://github.com/gravitational/teleport/pull/25247) and due to security concerns we likely do not want to allow the DB agent to have write permissions on `types.KindDatabase`